src/banman.h

```diff
@@ -60,40 +60,37 @@ class BanMan
 public:
     ~BanMan();
     BanMan(fs::path ban_file, CClientUIInterface* client_interface, int64_t default_ban_time);
-    void Ban(const CNetAddr& net_addr, int64_t ban_time_offset = 0, bool since_unix_epoch = false);
-    void Ban(const CSubNet& sub_net, int64_t ban_time_offset = 0, bool since_unix_epoch = false);
-    void Discourage(const CNetAddr& net_addr);
-    void ClearBanned();
+    void Ban(const CNetAddr& net_addr, int64_t ban_time_offset = 0, bool since_unix_epoch = false) EXCLUSIVE_LOCKS_REQUIRED(!m_banned_mutex);
+    void Ban(const CSubNet& sub_net, int64_t ban_time_offset = 0, bool since_unix_epoch = false) EXCLUSIVE_LOCKS_REQUIRED(!m_banned_mutex);
```

Comment:
I find it strange that there is a requirement to not hold the mutex when calling the constructor because at the time the constructor is called the mutex does not exist yet (it is a member variable).

I checked that if I remove these annotations, then indeed `-Wthread-safety-negative` produces a warning due to the `LOCK()` inside the constructor. The `LOCK()` is not needed either because the constructor will not be called concurrently with another method (or constructor). Removing the `LOCK()` of course causes warnings that touching `m_banned` requires the mutex :( I guess this is some weird limitation in the thread safety analysis.

Reply:
The constructor is `BanMan`, the annotated functions are `Ban` ?

Reply:
Right! :facepalm: